### PR TITLE
store: read postings directly into diffvarint format

### PR DESCRIPTION
**Pull request by GiedriusS**
Instead of allocating bytes for raw postings, let's read them directly into diff varint format to save memory. ~This should already cut memory usage by 3x.~ Savings are smaller. See comment.

### CODE REVIEW

File: pkg/store/postings.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	if err != nil {
+		return 0, errors.Wrap(err, "reading")
+	}
+	if read != 4 {
+		return 0, fmt.Errorf("read got %d bytes instead of 4", read)
+	}
+	return binary.BigEndian.Uint32(buf), nil
+}
+
+func (r *postingsReaderBuilder) Next() bool {
```
Comment (yeya24) [inline pkg/store/postings.go:60]: Can we check context as well?

File: pkg/store/postings.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+		}
+		r.lastOffset += 4
+
+		uvarintSize := binary.PutUvarint(r.uvarintEncodeBuf, uint64(posting-prev))
+		r.cur = append(r.cur, r.uvarintEncodeBuf[:uvarintSize]...)
+		prev = posting
+	}
+	r.numberOfPostingsInCur = uint64(postingsCount)
+
+	r.pi++
```
Comment (yeya24) [inline pkg/store/postings.go:110]: I guess we need to update `r.keyID`?

**Comment by yeya24** [comment]
Test failure seems related. If CI passed then I am good. It is a nice improvement.

**Comment by GiedriusS** [comment]
Last benchmarks:
```go
name                                                                      old time/op    new time/op    delta
BucketIndexReader_ExpandedPostings/n="1"-16                                 19.1ms ± 0%    14.2ms ± 2%  -25.74%  (p=0.000 n=8+10)
BucketIndexReader_ExpandedPostings/n="1",j="foo"-16                          113ms ± 1%      52ms ± 1%  -54.11%  (p=0.000 n=10+9)
BucketIndexReader_ExpandedPostings/j="foo",n="1"-16                          113ms ± 0%      53ms ± 2%  -53.34%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/n="1",j!="foo"-16                         106ms ± 1%      51ms ± 2%  -52.14%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/i=~".*"-16                                567ms ± 1%     422ms ± 2%  -25.61%  (p=0.000 n=10+9)
BucketIndexReader_ExpandedPostings/i=~".+"-16                                979ms ± 1%     899ms ± 4%   -8.24%  (p=0.000 n=9+10)
BucketIndexReader_ExpandedPostings/i=~""-16                                  1.02s ± 1%     0.72s ± 3%  -29.60%  (p=0.000 n=10+9)
BucketIndexReader_ExpandedPostings/i!=""-16                                  925ms ± 2%     832ms ± 3%  -10.03%  (p=0.000 n=10+9)
BucketIndexReader_ExpandedPostings/n="1",i=~".*",j="foo"-16                  183ms ± 1%     118ms ± 2%  -35.47%  (p=0.000 n=9+10)
BucketIndexReader_ExpandedPostings/n="1",i=~".*",i!="2",j="foo"-16           190ms ± 2%     120ms ± 1%  -36.91%  (p=0.000 n=9+8)
BucketIndexReader_ExpandedPostings/n="1",i!=""-16                            537ms ± 2%     456ms ± 8%  -15.05%  (p=0.000 n=9+10)
BucketIndexReader_ExpandedPostings/n="1",i!="",j="foo"-16                    544ms ± 1%     456ms ± 3%  -16.32%  (p=0.000 n=8+10)
BucketIndexReader_ExpandedPostings/n="1",i=~".+",j="foo"-16                  615ms ± 2%     524ms ± 6%  -14.82%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/n="1",i=~"1.+",j="foo"-16                 115ms ± 2%      64ms ± 7%  -44.60%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/n="1",i=~".+",i!="2",j="foo"-16           636ms ± 3%     531ms ± 3%  -16.42%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/n="1",i=~".+",i!~"2.*",j="foo"-16         704ms ± 2%     579ms ± 4%  -17.70%  (p=0.000 n=9+10)
BucketIndexReader_ExpandedPostings/i=~"0|1|2"-16                            68.3µs ± 1%    72.0µs ± 0%   +5.35%  (p=0.000 n=9+10)
BucketIndexReader_ExpandedPostings/uniq=~"9|random-shuffled-values|1"-16     180ms ± 0%     178ms ± 1%   -0.81%  (p=0.004 n=9+10)

name                                                                      old alloc/op   new alloc/op   delta
BucketIndexReader_ExpandedPostings/n="1"-16                                 9.88MB ± 0%    9.45MB ± 0%   -4.40%  (p=0.000 n=10+9)
BucketIndexReader_ExpandedPostings/n="1",j="foo"-16                         16.7MB ± 1%    12.1MB ± 3%  -27.67%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/j="foo",n="1"-16                         16.7MB ± 0%    12.0MB ± 1%  -28.38%  (p=0.000 n=8+8)
BucketIndexReader_ExpandedPostings/n="1",j!="foo"-16                        16.7MB ± 1%    12.1MB ± 2%  -27.87%  (p=0.000 n=9+10)
BucketIndexReader_ExpandedPostings/i=~".*"-16                                238MB ± 0%     226MB ± 0%   -5.02%  (p=0.000 n=8+9)
BucketIndexReader_ExpandedPostings/i=~".+"-16                                333MB ± 0%     319MB ± 0%   -4.02%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/i=~""-16                                  158MB ± 0%     134MB ± 0%  -15.11%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/i!=""-16                                  333MB ± 0%     320MB ± 0%   -3.96%  (p=0.000 n=9+10)
BucketIndexReader_ExpandedPostings/n="1",i=~".*",j="foo"-16                 18.3MB ± 1%    13.6MB ± 3%  -25.85%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/n="1",i=~".*",i!="2",j="foo"-16          19.7MB ± 1%    14.4MB ± 3%  -27.08%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/n="1",i!=""-16                            134MB ± 0%     121MB ± 0%   -9.88%  (p=0.000 n=9+9)
BucketIndexReader_ExpandedPostings/n="1",i!="",j="foo"-16                    141MB ± 0%     124MB ± 0%  -12.23%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/n="1",i=~".+",j="foo"-16                  141MB ± 0%     124MB ± 0%  -12.21%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/n="1",i=~"1.+",j="foo"-16                27.9MB ± 0%    21.5MB ± 2%  -23.04%  (p=0.000 n=9+10)
BucketIndexReader_ExpandedPostings/n="1",i=~".+",i!="2",j="foo"-16           142MB ± 0%     125MB ± 1%  -12.49%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/n="1",i=~".+",i!~"2.*",j="foo"-16         156MB ± 0%     139MB ± 0%  -11.20%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/i=~"0|1|2"-16                            14.4kB ± 1%    12.3kB ± 1%  -14.19%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/uniq=~"9|random-shuffled-values|1"-16     126MB ± 0%      49MB ± 0%  -61.62%  (p=0.000 n=10+10)

name                                                                      old allocs/op  new allocs/op  delta
BucketIndexReader_ExpandedPostings/n="1"-16                                    154 ± 1%       162 ± 2%   +4.72%  (p=0.000 n=8+10)
BucketIndexReader_ExpandedPostings/n="1",j="foo"-16                            427 ± 1%       484 ± 3%  +13.34%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/j="foo",n="1"-16                            427 ± 1%       480 ± 0%  +12.40%  (p=0.001 n=8+6)
BucketIndexReader_ExpandedPostings/n="1",j!="foo"-16                           425 ± 0%       480 ± 2%  +12.99%  (p=0.000 n=8+10)
BucketIndexReader_ExpandedPostings/i=~".*"-16                                1.23k ± 0%     1.30k ± 1%   +5.47%  (p=0.000 n=8+9)
BucketIndexReader_ExpandedPostings/i=~".+"-16                                1.30M ± 0%     1.30M ± 0%   -0.00%  (p=0.000 n=9+10)
BucketIndexReader_ExpandedPostings/i=~""-16                                  1.30M ± 0%     1.30M ± 0%   +0.01%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/i!=""-16                                  1.30M ± 0%     1.30M ± 0%   -0.00%  (p=0.020 n=10+10)
BucketIndexReader_ExpandedPostings/n="1",i=~".*",j="foo"-16                    434 ± 1%       486 ± 3%  +11.96%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/n="1",i=~".*",i!="2",j="foo"-16             513 ± 2%       562 ± 3%   +9.41%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/n="1",i!=""-16                            1.30M ± 0%     1.30M ± 0%     ~     (p=0.899 n=10+10)
BucketIndexReader_ExpandedPostings/n="1",i!="",j="foo"-16                    1.30M ± 0%     1.30M ± 0%   +0.00%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/n="1",i=~".+",j="foo"-16                  1.30M ± 0%     1.30M ± 0%   +0.00%  (p=0.000 n=10+9)
BucketIndexReader_ExpandedPostings/n="1",i=~"1.+",j="foo"-16                  145k ± 0%      145k ± 0%   +0.02%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/n="1",i=~".+",i!="2",j="foo"-16           1.30M ± 0%     1.30M ± 0%   +0.00%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/n="1",i=~".+",i!~"2.*",j="foo"-16         1.45M ± 0%     1.43M ± 0%   -0.77%  (p=0.000 n=10+9)
BucketIndexReader_ExpandedPostings/i=~"0|1|2"-16                               144 ± 0%       150 ± 0%   +4.17%  (p=0.000 n=10+10)
BucketIndexReader_ExpandedPostings/uniq=~"9|random-shuffled-values|1"-16      750k ± 0%      800k ± 0%   +6.67%  (p=0.000 n=10+10)
```

`i=~"0|1|2"` case of increased CPU usage is interesting. We could merge this and I'll investigate further.

**Comment by yeya24** [comment]
QQ @GiedriusS, this doesn't solve https://github.com/thanos-io/thanos/issues/6434, right?

**Comment by GiedriusS** [comment]
No, it's a different problem. I'll try to tackle #6434 next

File: pkg/store/bucket.go [hunk 3]
```diff
@@ -2570,52 +2576,60 @@ func (r *bucketIndexReader) fetchPostings(ctx context.Context, keys []labels.Lab
...
+				r.stats.PostingsTouchedSizeSum += units.Base2Bytes(int(len(diffVarintPostings)))
 				r.stats.cachedPostingsCompressions += 1
-				r.stats.cachedPostingsCompressionErrors += compressionErrors
-				r.stats.CachedPostingsOriginalSizeSum += units.Base2Bytes(len(pBytes))
-				r.stats.CachedPostingsCompressedSizeSum += units.Base2Bytes(compressedSize)
-				r.stats.CachedPostingsCompressionTimeSum += compressionTime
+				r.stats.CachedPostingsOriginalSizeSum += units.Base2Bytes(len(diffVarintPostings))
+				r.stats.CachedPostingsCompressedSizeSum += units.Base2Bytes(len(dataToCache))
+				r.stats.CachedPostingsCompressionTimeSum += time.Since(startCompression)
 				r.mtx.Unlock()
```
Comment (yeya24) [inline pkg/store/bucket.go:2622]: I am wondering if it is better to hold the lock only once and add to stats at the end? But it is a small nit

**Review by yeya24 — APPROVED** [review]

File: pkg/store/bucket.go (hunk 3 above)
Comment (yeya24) [reply pkg/store/bucket.go:2622]: If it is really an issue, maybe we can change next pr

**Pull request merged**